File: Mancala.cpp
#include <iostream>
#include <vector>
#include <string>

#define MAX_SEEDS 999

void print_board(const std::vector<int>& board, int number_of_pits, int total_pits);
bool make_move(std::vector<int>& board, int number_of_pits, int total_pits, int player, int pit);
bool check_game_over(std::vector<int>& board, int number_of_pits, int total_pits);
void handle_exit();
void print_rules();
void print_info();

int main() {
    int number_of_pits, seeds_per_pit;
    std::string input;

    std::cout << "Type 'rule' to view the rules, 'info' to view the information, or 'exit' to quit anytime.\n\n";

    // Get the number of pits from the user
    while (true) {
        std::cout << "Enter the number of pits each player has: ";
        std::getline(std::cin, input);
        if (input == "exit") {
            handle_exit();
            continue;
        } else if (input == "rule") {
            print_rules();
            continue;
        } else if (input == "info") {
            print_info();
            continue;
        }
        try {
            number_of_pits = std::stoi(input);
            if (number_of_pits > 0) break;
            else std::cout << "Invalid input. Please enter a positive integer.\n";
        } catch (std::exception& e) {
            std::cout << "Invalid input. Please enter a positive integer.\n";
        }
    }

    // Get the number of seeds per pit from the user
    while (true) {
        std::cout << "Enter the number of seeds per pit: ";
        std::getline(std::cin, input);
        if (input == "exit") {
            handle_exit();
            continue;
        } else if (input == "rule") {
            print_rules();
            continue;
        } else if (input == "info") {
            print_info();
            continue;
        }
        try {
            seeds_per_pit = std::stoi(input);
            if (seeds_per_pit > 0 && seeds_per_pit * number_of_pits <= MAX_SEEDS) break;
            else std::cout << "Invalid input. Please enter a positive integer between 1 and " << MAX_SEEDS / number_of_pits << ".\n";
        } catch (std::exception& e) {
            std::cout << "Invalid input. Please enter a positive integer between 1 and " << MAX_SEEDS / number_of_pits << ".\n";
        }
    }

    int total_pits = number_of_pits * 2 + 2; // 2 stores plus the pits
    std::vector<int> board(total_pits, seeds_per_pit);
    board[number_of_pits] = 0; // Player 1's store
    board[total_pits - 1] = 0; // Player 2's store

    // Explain the board setup
    std::cout << "\nGame setup:\n";
    std::cout << "  - Player 1's pits start on the left side of the board.\n";
    std::cout << "  - Player 2's pits start on the right side of the board.\n";
    std::cout << "  - Player 1's store is at the right end of the board.\n";
    std::cout << "  - Player 2's store is at the left end of the board.\n";
    std::cout << "  - The pits and stores are arranged in a line, with Player 1's pits followed by Player 1's store, then Player 2's pits and Player 2's store.\n";
    std::cout << "  - Each player has " << number_of_pits << " pits and " << seeds_per_pit << " seeds in each pit initially.\n";
    std::cout << "Type 'rule' to view the rules, 'info' to view the information, or 'exit' to quit anytime.\n\n";

    int player = 0;
    int pit;
    bool game_over = false;

    while (!game_over) {
        print_board(board, number_of_pits, total_pits);
        
        // Player makes a move
        std::cout << "Player " << player + 1 << ", choose a pit (1-" << number_of_pits << ") : ";
        std::getline(std::cin, input);

        if (input == "exit") {
            handle_exit();
            continue;
        } else if (input == "rule") {
            print_rules();
            print_board(board, number_of_pits, total_pits);
            continue;
        } else if (input == "info") {
            print_info();
            print_board(board, number_of_pits, total_pits);
            continue;
        }

        try {
            pit = std::stoi(input) - 1; // Adjust for 0-indexing
            if (pit < 0 || pit >= number_of_pits) {
                std::cout << "Invalid pit number. Try again.\n";
                continue;
            }

            bool another_turn = make_move(board, number_of_pits, total_pits, player, pit);
            game_over = check_game_over(board, number_of_pits, total_pits);
            
            // Switch players if the last seed didn't land in the player's store
            if (!game_over && !another_turn) {
                player = (player + 1) % 2;
            }
        } catch (std::exception& e) {
            std::cout << "Invalid input. Please enter a pit number, 'exit', or 'rule'.\n";
        }
    }
    
    // Determine the winner
    int player1_store = board[number_of_pits];
    int player2_store = board[total_pits - 1];
    
    if (player1_store > player2_store) {
        std::cout << "Player 1 wins with " << player1_store << " seeds!\n";
    } else if (player2_store > player1_store) {
        std::cout << "Player 2 wins with " << player2_store << " seeds!\n";
    } else {
        std::cout << "It's a tie!\n";
    }
    print_board(board, number_of_pits, total_pits);

    return 0;
}

void print_board(const std::vector<int>& board, int number_of_pits, int total_pits) {
    std::cout << "\nPlayer 2\n";
    std::cout << " ";
    for (int i = total_pits - 2; i > number_of_pits; i--) {
        std::cout << " " << board[i] << " ";
    }
    std::cout << "\n " << board[total_pits - 1];
    for (int i = 0; i < number_of_pits + 1; i++) {
        std::cout << "  ";
    }
    std::cout << "   " << board[number_of_pits] << "\n ";
    for (int i = 0; i < number_of_pits; i++) {
        std::cout << " " << board[i] << " ";
    }
    std::cout << "\nPlayer 1\n\n";
}

bool make_move(std::vector<int>& board, int number_of_pits, int total_pits, int player, int pit) {
    int start = player == 0 ? 0 : number_of_pits + 1;
    int pos = start + pit;
    int seeds = board[pos];

    if (seeds == 0) return true; // Move again if the pit is empty

    board[pos] = 0;
    
    while (seeds > 0) {
        pos = (pos + 1) % total_pits;
        if (pos == (player == 0 ? total_pits - 1 : number_of_pits)) continue; // Skip opponent's store
        board[pos]++;
        seeds--;
    }

    // Check for capture
    if (pos != number_of_pits && pos != total_pits - 1) {  // Last seed not in a store
        if (player == pos / (number_of_pits + 1) && board[pos] == 1) {  // Last seed in player's empty pit
            int opposite = total_pits - 2 - pos;
            if (board[opposite] > 0) {
                board[player == 0 ? number_of_pits : total_pits - 1] += board[opposite] + 1;
                board[opposite] = 0;
                board[pos] = 0;
            }
        }
    }

    // Return true if the last seed was placed in the player's store
    return (player == 0 && pos == number_of_pits) || (player == 1 && pos == total_pits - 1);
}

bool check_game_over(std::vector<int>& board, int number_of_pits, int total_pits) {
    int sum1 = 0, sum2 = 0;
    for (int i = 0; i < number_of_pits; i++) {
        sum1 += board[i];
        sum2 += board[number_of_pits + 1 + i];
    }

    if (sum1 == 0) {
        for (int i = 0; i < number_of_pits; i++) {
            board[number_of_pits] += board[i];
            board[i] = 0;
        }
        return true;
    }

    if (sum2 == 0) {
        for (int i = number_of_pits + 1; i < total_pits; i++) {
            board[total_pits - 1] += board[i];
            board[i] = 0;
        }
        return true;
    }

    return false;
}

void handle_exit() {
    std::string choice;
    
    std::cout << "Are you sure you want to quit? (y/n): ";
    std::getline(std::cin, choice);

    if (std::cin.fail()) {
        std::cout << "\n";
        return;
    }
    else if (choice.substr(0, 1) == "y" || choice.substr(0, 1) == "Y") {
        std::cout << "Exiting the game...\n";
        exit(0);
    } else {
        std::cout << "\n";
        return;
    }
}

void print_rules() {
    std::cout << "\nMancala Rules:\n";
    std::cout << "1. The board has two rows of pits, one for each player. Each player has a row of pits and a store.\n";
    std::cout << "2. Each player has a certain number of pits (defined at the beginning of the game) and each pit starts with the same number of seeds (also defined at the beginning).\n";
    std::cout << "3. The board setup is as follows:\n";
    std::cout << "  - Player 1's pits start on the left side of the board.\n";
    std::cout << "  - Player 2's pits start on the right side of the board.\n";
    std::cout << "  - Player 1's store is at the right end of the board.\n";
    std::cout << "  - Player 2's store is at the left end of the board.\n";
    std::cout << "  - The pits and stores are arranged in a line, with Player 1's pits followed by Player 1's store, then Player 2's pits and Player 2's store.\n";
    std::cout << "4. Players take turns selecting one of their pits to move the seeds. The seeds are distributed counter-clockwise around the board.\n";
    std::cout << "5. Seeds are placed one by one in each subsequent pit, including the player's store but excluding the opponent's store.\n";
    std::cout << "6. If the last seed lands in the player's store, they get another turn.\n";
    std::cout << "7. If the last seed lands in an empty pit on the player's side, they capture all seeds in the opposite pit (if it contains seeds) and add them to their store.\n";
    std::cout << "8. The game ends when all pits on one side are empty. The remaining seeds are moved to the respective stores.\n";
    std::cout << "9. The player with the most seeds in their store at the end of the game wins.\n";
    std::cout << "10. If both players have the same number of seeds in their stores, the game is a tie.\n";
    std::cout << "Type 'rule' to view the rules, 'info' to view the information, or 'exit' to quit anytime.\n\n";
}

void print_info() {
    std::cout << "\nThis is Mancala made by `https://github.com/Willie169`.\n\n";
}

File: Mancala.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Mancala {
    private static final int MAX_SEEDS = 999;

    public static void main(String[] args) {
        int numberOfPits, seedsPerPit;
        String input;
        Scanner scanner = new Scanner(System.in);

        System.out.print("Type 'rule' to view the rules, 'info' to view the information, or 'exit' to quit anytime.\n\n");

        // Get the number of pits from the user
        while (true) {
            System.out.print("Enter the number of pits each player has: ");
            input = scanner.nextLine();
            if (input.equals("exit")) {
                handleExit();
                continue;
            } else if (input.equals("rule")) {
                printRules();
                continue;
            } else if (input.equals("info")) {
                printInfo();
                continue;
            }
            try {
                numberOfPits = Integer.parseInt(input);
                if (numberOfPits > 0) break;
                else System.out.print("Invalid input. Please enter a positive integer.\n");
            } catch (NumberFormatException e) {
                System.out.print("Invalid input. Please enter a positive integer.\n");
            }
        }

        // Get the number of seeds per pit from the user
        while (true) {
            System.out.print("Enter the number of seeds per pit: ");
            input = scanner.nextLine();
            if (input.equals("exit")) {
                handleExit();
                continue;
            } else if (input.equals("rule")) {
                printRules();
                continue;
            } else if (input.equals("info")) {
                printInfo();
                continue;
            }
            try {
                seedsPerPit = Integer.parseInt(input);
                if (seedsPerPit > 0 && seedsPerPit * numberOfPits <= MAX_SEEDS) break;
                else System.out.print("Invalid input. Please enter a positive integer between 1 and " + MAX_SEEDS / numberOfPits + ".\n");
            } catch (NumberFormatException e) {
                System.out.print("Invalid input. Please enter a positive integer between 1 and " + MAX_SEEDS / numberOfPits + ".\n");
            }
        }

        int totalPits = numberOfPits * 2 + 2; // 2 stores plus the pits
        List<Integer> board = new ArrayList<>(totalPits);
        for (int i = 0; i < totalPits; i++) {
            board.add(seedsPerPit);
        }
        board.set(numberOfPits, 0); // Player 1's store
        board.set(totalPits - 1, 0); // Player 2's store

        // Explain the board setup
        System.out.print("\nGame setup:\n");
        System.out.print("  - Player 1's pits start on the left side of the board.\n");
        System.out.print("  - Player 2's pits start on the right side of the board.\n");
        System.out.print("  - Player 1's store is at the right end of the board.\n");
        System.out.print("  - Player 2's store is at the left end of the board.\n");
        System.out.print("  - The pits and stores are arranged in a line, with Player 1's pits followed by Player 1's store, then Player 2's pits and Player 2's store.\n");
        System.out.print("  - Each player has " + numberOfPits + " pits and " + seedsPerPit + " seeds in each pit initially.\n");
        System.out.print("Type 'rule' to view the rules, 'info' to view the information, or 'exit' to quit anytime.\n\n");

        int player = 0;
        int pit;
        boolean gameOver = false;

        while (!gameOver) {
            printBoard(board, numberOfPits, totalPits);

            // Player makes a move
            System.out.print("Player " + (player + 1) + ", choose a pit (1-" + numberOfPits + ") : ");
            input = scanner.nextLine();

            if (input.equals("exit")) {
                handleExit();
                continue;
            } else if (input.equals("rule")) {
                printRules();
                printBoard(board, numberOfPits, totalPits);
                continue;
            } else if (input.equals("info")) {
                printInfo();
                printBoard(board, numberOfPits, totalPits);
                continue;
            }

            try {
                pit = Integer.parseInt(input) - 1; // Adjust for 0-indexing
                if (pit < 0 || pit >= numberOfPits) {
                    System.out.print("Invalid pit number. Try again.\n");
                    continue;
                }

                boolean anotherTurn = makeMove(board, numberOfPits, totalPits, player, pit);
                gameOver = checkGameOver(board, numberOfPits, totalPits);

                // Switch players if the last seed didn't land in the player's store
                if (!gameOver && !anotherTurn) {
                    player = (player + 1) % 2;
                }
            } catch (NumberFormatException e) {
                System.out.print("Invalid input. Please enter a pit number, 'exit', or 'rule'.\n");
            }
        }

        // Determine the winner
        int player1Store = board.get(numberOfPits);
        int player2Store = board.get(totalPits - 1);

        if (player1Store > player2Store) {
            System.out.print("Player 1 wins with " + player1Store + " seeds!\n");
        } else if (player2Store > player1Store) {
            System.out.print("Player 2 wins with " + player2Store + " seeds!\n");
        } else {
            System.out.print("It's a tie!\n");
        }
        printBoard(board, numberOfPits, totalPits);
    }

    private static void printBoard(List<Integer> board, int numberOfPits, int totalPits) {
        System.out.print("\nPlayer 2\n");
        System.out.print(" ");
        for (int i = totalPits - 2; i > numberOfPits; i--) {
            System.out.print(" " + board.get(i) + " ");
        }
        System.out.print("\n " + board.get(totalPits - 1));
        for (int i = 0; i < numberOfPits + 1; i++) {
            System.out.print("  ");
        }
        System.out.print("   " + board.get(numberOfPits) + "\n ");
        for (int i = 0; i < numberOfPits; i++) {
            System.out.print(" " + board.get(i) + " ");
        }
        System.out.print("\nPlayer 1\n\n");
    }

    private static boolean makeMove(List<Integer> board, int numberOfPits, int totalPits, int player, int pit) {
        int start = player == 0 ? 0 : numberOfPits + 1;
        int pos = start + pit;
        int seeds = board.get(pos);

        if (seeds == 0) return true; // Move again if the pit is empty

        board.set(pos, 0);

        while (seeds > 0) {
            pos = (pos + 1) % totalPits;
            if (pos == (player == 0 ? totalPits - 1 : numberOfPits)) continue; // Skip opponent's store
            board.set(pos, board.get(pos) + 1);
            seeds--;
        }

        // Check for capture
        if (pos != numberOfPits && pos != totalPits - 1) {  // Last seed not in a store
            if (player == pos / (numberOfPits + 1) && board.get(pos) == 1) {  // Last seed in player's empty pit
                int opposite = totalPits - 2 - pos;
                if (board.get(opposite) > 0) {
                    board.set(player == 0 ? numberOfPits : totalPits - 1, board.get(player == 0 ? numberOfPits : totalPits - 1) + board.get(opposite) + 1);
                    board.set(opposite, 0);
                    board.set(pos, 0);
                }
            }
        }

        // Return true if the last seed was placed in the player's store
        return (player == 0 && pos == numberOfPits) || (player == 1 && pos == totalPits - 1);
    }

    private static boolean checkGameOver(List<Integer> board, int numberOfPits, int totalPits) {
        int sum1 = 0, sum2 = 0;
        for (int i = 0; i < numberOfPits; i++) {
            sum1 += board.get(i);
            sum2 += board.get(numberOfPits + 1 + i);
        }

        if (sum1 == 0) {
            for (int i = 0; i < numberOfPits; i++) {
                board.set(numberOfPits, board.get(numberOfPits) + board.get(i));
                board.set(i, 0);
            }
            return true;
        }

        if (sum2 == 0) {
            for (int i = numberOfPits + 1; i < totalPits; i++) {
                board.set(totalPits - 1, board.get(totalPits - 1) + board.get(i));
                board.set(i, 0);
            }
            return true;
        }

        return false;
    }

    private static void handleExit() {
        Scanner scanner = new Scanner(System.in);
        String choice;

        System.out.print("Are you sure you want to quit? (y/n): ");
        choice = scanner.nextLine();

        if (choice.substring(0, 1).equalsIgnoreCase("y")) {
            System.out.print("Exiting the game...\n");
            System.exit(0);
        } else {
            System.out.print("\n");
        }
    }

    private static void printRules() {
        System.out.print("\nMancala Rules:\n");
        System.out.print("1. The board has two rows of pits, one for each player. Each player has a row of pits and a store.\n");
        System.out.print("2. Each player has a certain number of pits (defined at the beginning of the game) and each pit starts with the same number of seeds (also defined at the beginning).\n");
        System.out.print("3. The board setup is as follows:\n");
        System.out.print("  - Player 1's pits start on the left side of the board.\n");
        System.out.print("  - Player 2's pits start on the right side of the board.\n");
        System.out.print("  - Player 1's store is at the right end of the board.\n");
        System.out.print("  - Player 2's store is at the left end of the board.\n");
        System.out.print("  - The pits and stores are arranged in a line, with Player 1's pits followed by Player 1's store, then Player 2's pits and Player 2's store.\n");
        System.out.print("4. Players take turns selecting one of their pits to move the seeds. The seeds are distributed counter-clockwise around the board.\n");
        System.out.print("5. Seeds are placed one by one in each subsequent pit, including the player's store but excluding the opponent's store.\n");
        System.out.print("6. If the last seed lands in the player's store, they get another turn.\n");
        System.out.print("7. If the last seed lands in an empty pit on the player's side, they capture all seeds in the opposite pit (if it contains seeds) and add them to their store.\n");
        System.out.print("8. The game ends when all pits on one side are empty. The remaining seeds are moved to the respective stores.\n");
        System.out.print("9. The player with the most seeds in their store at the end of the game wins.\n");
        System.out.print("10. If both players have the same number of seeds in their stores, the game is a tie.\n");
        System.out.print("Type 'rule' to view the rules, 'info' to view the information, or 'exit' to quit anytime.\n\n");
    }

    private static void printInfo() {
        System.out.print("\nThis is Mancala made by `https://github.com/Willie169`.\n\n");
    }
}
